### DISCUSSION TITLE
Full Topic List not appearing on RPi 4

### DISCUSSION THREAD

**Discussion by ghost** [iRobotEducation/create3_docs#77] [Questions]
Hello!

I'm very excited to begin using the Create3 for some interesting SLAM projects. I've spent the better part of two days trying to get the basic ros2 setup running on a RPi 4 that I have connected via usb to my Create 3.

I follow the instructions on your website: https://iroboteducation.github.io/create3_docs/setup/pi4/

However, after completing them and checking for the available topics, All I see is

```
ubuntu@irobot:~$ ros2 topic list
/parameter_events
/rosout
```

Please help me figure out what's wrong with my installation.

**System Info:**
Ubuntu 20.04.4 LTS (GNU/Linux 5.4.0-1062-raspi aarch64)
RMW_IMPLEMENTATION=rmw_cyclonedds_cpp

Note: I also tried the RWM Fast-DDS since I didn't understand in the instructions where to find my robot's "Application Configuration page"

**Bottom of `~/.bashrc`:**
```
export _colcon_cd_root=/opt/ros/galactic/
source /usr/share/colcon_argcomplete/hook/colcon-argcomplete.bash
source /opt/ros/galactic/setup.bash
export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp
export CYCLONEDDS_URI=/home/ubuntu/ros_install/profile.xml
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/ubuntu/ros_install/profile_fast.xml
```

**RMW profile (located in ~/ros_install/):**
```
<CycloneDDS>
   <Domain>
     <General>
        <DontRoute>true</DontRoute>
    </General>
   </Domain>
</CycloneDDS>
```

This profile was found [here](https://iroboteducation.github.io/create3_docs/setup/xml-config/) under Multiple network Interfaces due to being on my company's WiFi

**Comment by shamlian** [comment]
Hi there,

I'm a little confused. It looks like that RMW profile you're using is one you would want on a laptop; on your Pi, if you're connected both to the robot with usb0 and to your local network with wlan0, you'll want to use the multiple network interfaces XML config on the Pi, which is

```
<CycloneDDS>
   <Domain>
     <General>
        <NetworkInterfaceAddress>usb0,wlan0</NetworkInterfaceAddress>
    </General>
   </Domain>
</CycloneDDS>
```

Using this, you _should_ be able to see the robot with a ros2 topic list. The robot's RMW can be forced to Cyclone by connecting to it with a browser to {robot ip}/ros-config, or if you know you're connected over usb0 with the Pi (can you ping 192.168.186.2?), you can just  `curl -X POST -d "ros_domain_id=0&ros_namespace=&rmw_implementation=rmw_cyclonedds_cpp" "http://192.168.186.2/ros-config-save-main"` from the prompt.

You will need to restart both the robot and also ROS 2 on your Pi (`ros2 daemon stop && ros2 topic list` -- daemon will start automatically) for configuration changes to take effect. (Technically, you only need to restart the application on the robot, but if you don't have access to the web interface, that might be more of a challenge than it's worth -- easier just to fully power the robot down and then place it on the dock to boot up.)

**Reply by ghost** [reply]
Hi shamlian, thank you for the correction! I misread the text in the Attention! section of https://iroboteducation.github.io/create3_docs/setup/xml-config/ and so used the wrong one. 

However I am unable to reach the {robot_ip}/ros-config via browser.

Pinging 192.168.186.2 was successful:
```
ubuntu@irobot:~$ ping 192.168.186.2
PING 192.168.186.2 (192.168.186.2) 56(84) bytes of data.
64 bytes from 192.168.186.2: icmp_seq=1 ttl=64 time=0.512 ms
64 bytes from 192.168.186.2: icmp_seq=2 ttl=64 time=0.533 ms
```

but the curl command hits un implemented method:
```
ubuntu@irobot:~$ curl -X POST -d "ros_domain_id=0&ros_namespace=&rmw_implementation=rmw_cyclonedds_cpp" "http://192.168.186.2/ros-config-save-main"
<HTML><HEAD><TITLE>501 Not Implemented</TITLE></HEAD>
<BODY><H1>501 Not Implemented</H1>
The requested method is not recognized
</BODY></HTML>
```

For the case where my system was correctly on rmw_cyclonedds_cpp already, I also restarted the robot via `sudo reboot` and then ran the provided command `ros2 daemon stop && ros2 topic list`.
```
ubuntu@irobot:~$ ros2 daemon stop && ros2 topic list
The daemon is not running
/parameter_events
/rosout
ubuntu@irobot:~$ ros2 daemon stop && ros2 topic list
The daemon has been stopped
/parameter_events
/rosout
```

Unfortunately none of those worked.


Additional Info:
```
ubuntu@irobot:~$ ros2 doctor
All 4 checks passed
```

**Reply by shamlian** [reply]
Have you updated your robot to G.2.2 as in the directions at https://edu.irobot.com/create3-setup ? Also -- did you mean that you restarted your *Pi* with `sudo reboot`? The robot does not allow shell access.

Details on changing the RMW from the robot's web server are [here](https://iroboteducation.github.io/create3_docs/setup/provision/).

**Reply by ghost** [reply]
Aha! I think you may have found the answer to my problem, I didn't realize it needed to be updated. I'm current working on this remotely and so can only interface with the pi but will attempt this fix on Monday and update this discussion with the results. Thanks for the help thus far!

**Reply by shamlian** [reply]
You're welcome! Have a great weekend.

**Reply by ghost** [reply]
Hello again, Shamlian.

I was able to go through the process of upgrading the firmware and checking the Application page of the robot to confirm it was configured for CycloneDDS RMW. However when I went to restart the robot and ROS2 server on the pi, I had issues restarting ROS2 and checking the topic list:

![image](https://user-images.githubusercontent.com/92594844/173435440-107e443f-fa8a-4acb-8c71-77cd1f34833a.png)

**Reply by shamlian** [reply]
Good morning. It looks like there's a problem with your ROS 2 install and not with the robot. A couple questions:
1) Just to be sure, is the adapter board switch on the robot set to "USB" and not "BLE?" I think you said you could ping the robot from the Pi, but it is weird that you are getting "ros2: usb0: does not match an available interface supporting udp"
2) Can you give us the output of `ip a` from your Ubuntu machine?
3) As a sanity check, could you change the rmw on the robot and the Pi to Fast-DDS (robot: from the webserver choose rmw_fastrtps_cpp and then restart; pi: change RMW_IMPLEMENTATION to `rmw_fastrtps_cpp`) and remove the profile you added to .bashrc (`export FASTRTPS_DEFAULT_PROFILES_FILE=/home/ubuntu/ros_install/profile_fast.xml` -- delete the line)? Cyclone seems to have more trouble with multi-interface situations, so it's easier to debug basic connectivity with Fast, IMHO.
